# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [69]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [70]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143
The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! yo

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [71]:
import problem_unittests as tests
from string import punctuation
from collections import Counter

def create_lookup_tables(text):
#    print(text[:2000])
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    all_text = ' '.join([word for word in text])                      # consolidate to one string
#    all_text = ''.join([c for c in all_text if c not in punctuation]) # remove punctuation
#    all_text = all_text.lower()                                       # change upper to lower
    print(all_text[:2000]) # test

    text_split = all_text.split('\n')
    all_text = ''.join(text_split)

    words = all_text.split()

    word_counts = Counter(words)
    sorted_vocab = sorted(word_counts, key=word_counts.get, reverse=True)
    int_to_vocab = {ii: word for ii, word in enumerate(sorted_vocab, 0)}
#    int_to_vocab = {ii: word for ii, word in enumerate(sorted_vocab, 1)}

#    print(type(sorted_vocab)) # <class 'list'>
#    print(len(sorted_vocab))  # 71
#    print(type(int_to_vocab)) # <class 'dict'>
#    print(len(int_to_vocab))  # 71
    vocab_to_int = {word: ii for ii, word in int_to_vocab.items()}

#    for i in range(len(int_to_vocab)):
#        print(i)
#        print(int_to_vocab[i])
#    print(vocab_to_int['||Comma||'])
    
# return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

moe_szyslak moe's tavern where the elite meet to drink bart_simpson eh yeah hello is mike there last name rotch moe_szyslak hold on i'll check mike rotch mike rotch hey has anybody seen mike rotch lately moe_szyslak listen you little puke one of these days i'm gonna catch you and i'm gonna carve my name on your back with an ice pick moe_szyslak whats the matter homer you're not your normal effervescent self homer_simpson i got my problems moe give me another one moe_szyslak homer hey you should not drink to forget your problems barney_gumble yeah you should only drink to enhance your social skills
Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [72]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    punc_to_token = {
        '.': "||Period||", 
        ',': "||Comma||", 
        '"': "||Quotation_Mark||", 
        ';': "||Semicolon||", 
        '!': "||Exclamation_mark||", 
        '?': "||Question_mark||", 
        '(': "||Left_Parentheses||", 
        ')': "||Right_Parentheses||", 
        '-': "||Dash||", 
        '\n': "||Return||"
    }
    
    return punc_to_token

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to look at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [73]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

this is out ||period|| ||period|| ||period|| and out is one of the single most enjoyable experiences of life ||period|| people ||period|| ||period|| ||period|| did you ever hear people talking about we should go out ||question_mark|| this is what theyre talking about ||period|| ||period|| ||period|| this whole thing ||comma|| were all out now ||comma|| no one is home ||period|| not one person here is home ||comma|| were all out ||exclamation_mark|| there are people trying to find us ||comma|| they dont know where we are ||period|| ||left_parentheses|| on an imaginary phone ||right_parentheses|| did you ring ||question_mark|| ||comma|| i cant find him ||period|| where did he go ||question_mark|| he didnt tell me where he was going ||period|| he must have gone out ||period|| you wanna go out you get ready ||comma|| you pick out the clothes ||comma|| right ||question_mark|| you take the shower ||comma|| you get all ready ||comma|| get the cash ||comma|| get your friends ||comma|| the car 

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [74]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [75]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [76]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    features = np.zeros(((len(words) - sequence_length), sequence_length), dtype=int)
    targets = np.zeros((len(words) - sequence_length), dtype=int)
    for i in range(len(words) - sequence_length):
        features[i] = words[i : i + sequence_length]
        targets[i] = words[i + sequence_length]

## test
#    print(type(features)) # <class 'numpy.ndarray'>
#    print(features[0])    # [0 1 2 3 4]
#    print(targets[0])     # 5
#    print(features[1])    # [1 2 3 4 5]
#    print(targets[1])     # 6
#    print(features[len(words) - sequence_length - 1]) # [44 45 46 47 48]
#    print(targets[len(words) - sequence_length - 1])  # 49

    data = TensorDataset(torch.from_numpy(features), torch.from_numpy(targets))
    data_loader = torch.utils.data.DataLoader(data, shuffle=True, batch_size=batch_size)

    # return a dataloader
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own

## test
#test_text = range(50)
#batch_data(test_text, sequence_length=5, batch_size=10)

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [77]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[15, 16, 17, 18, 19],
        [24, 25, 26, 27, 28],
        [ 5,  6,  7,  8,  9],
        [44, 45, 46, 47, 48],
        [22, 23, 24, 25, 26],
        [ 3,  4,  5,  6,  7],
        [21, 22, 23, 24, 25],
        [38, 39, 40, 41, 42],
        [34, 35, 36, 37, 38],
        [ 7,  8,  9, 10, 11]])

torch.Size([10])
tensor([20, 29, 10, 49, 27,  8, 26, 43, 39, 12])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [78]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        
        # TODO: Implement function
        
        # set class variables
        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim

        # define model layers
        self.embd = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout = dropout, batch_first = True)
        self.dropout = nn.Dropout(0.3)
        self.fc = nn.Linear(hidden_dim, output_size)

    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function   
#        print(type(nn_input))
#        print(type(nn_input.size))
        batch_size = nn_input.size(0)
#        print(batch_size) # 50

        # embeddings and lstm_out
        embeds = self.embd(nn_input)
        lstm_out, hidden = self.lstm(embeds, hidden)
#        print(type(hidden))
        
        # stack up lstm outputs
        lstm_out = lstm_out.contiguous().view(-1, self.hidden_dim)

        # dropout and fully-connected layer
        out = self.dropout(lstm_out)
        out = self.fc(out)

        # reshape to be batch_size first
        out = out.view(batch_size, -1, self.output_size)
        out = out[:, -1] # get last batch of outputs

        # return one batch of output word scores and the hidden state
        return out, hidden
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        weight = next(self.parameters()).data

        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())

        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [79]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    clip=5 # gradient clipping
    
    # move data to GPU, if available
    if(train_on_gpu):
        inp, target = inp.cuda(), target.cuda()
    
    # perform backpropagation and optimization
    hidden = tuple([each.data for each in hidden])

    # zero accumulated gradients
    rnn.zero_grad()

    # get the output from the model
    output, hidden = rnn(inp, hidden)

    # calculate the loss and perform backprop
    loss = criterion(output.squeeze(), target)
    loss.backward()
    
    # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
    nn.utils.clip_grad_norm_(rnn.parameters(), clip)
    optimizer.step()

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [80]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
#                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
#                    epoch_i, n_epochs, np.average(batch_losses)))
                print('Epoch: {:>4}/{:<4}  Loss: {}'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of unique tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [81]:
# Data params
# Sequence Length
sequence_length = 10  # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [87]:
# Training parameters
# Number of Epochs
num_epochs = 50
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 256
# Hidden Dimension
hidden_dim = 800
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 640

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [88]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 50 epoch(s)...
Epoch:    1/50    Loss: 5.253700656816363
Epoch:    1/50    Loss: 4.782527927309275
Epoch:    1/50    Loss: 4.629239576682449
Epoch:    1/50    Loss: 4.565429873764515
Epoch:    1/50    Loss: 4.454377238079905
Epoch:    1/50    Loss: 4.429828142374754
Epoch:    1/50    Loss: 4.383285737410188
Epoch:    1/50    Loss: 4.351274823397398
Epoch:    1/50    Loss: 4.319598229974508
Epoch:    1/50    Loss: 4.3010002840310335
Epoch:    1/50    Loss: 4.300960117578507
Epoch:    1/50    Loss: 4.22971854172647
Epoch:    1/50    Loss: 4.243915567919612
Epoch:    1/50    Loss: 4.201597314327955
Epoch:    1/50    Loss: 4.236023548990488
Epoch:    1/50    Loss: 4.206151943653822
Epoch:    1/50    Loss: 4.153445649892092
Epoch:    1/50    Loss: 4.19222059212625
Epoch:    1/50    Loss: 4.207992108166218
Epoch:    1/50    Loss: 4.163940661028027
Epoch:    1/50    Loss: 4.169988998025656
Epoch:    2/50    Loss: 4.067798764377021
Epoch:    2/50    Loss: 4.020531105622649
Epoch: 

Epoch:   10/50    Loss: 3.619445611163974
Epoch:   10/50    Loss: 3.6026006851345302
Epoch:   10/50    Loss: 3.610572674870491
Epoch:   10/50    Loss: 3.6396705228835344
Epoch:   10/50    Loss: 3.6361037462949755
Epoch:   10/50    Loss: 3.650606483593583
Epoch:   10/50    Loss: 3.640276228263974
Epoch:   10/50    Loss: 3.6797466304153206
Epoch:   10/50    Loss: 3.6822531159967182
Epoch:   10/50    Loss: 3.6694066733121873
Epoch:   10/50    Loss: 3.6865705378353595
Epoch:   10/50    Loss: 3.679211489483714
Epoch:   10/50    Loss: 3.7019963815808294
Epoch:   10/50    Loss: 3.669753336906433
Epoch:   10/50    Loss: 3.697179463505745
Epoch:   10/50    Loss: 3.7272456292063
Epoch:   10/50    Loss: 3.7291508965194224
Epoch:   11/50    Loss: 3.6450283399687406
Epoch:   11/50    Loss: 3.578677800670266
Epoch:   11/50    Loss: 3.574112347513437
Epoch:   11/50    Loss: 3.5729835622012613
Epoch:   11/50    Loss: 3.6113914545625447
Epoch:   11/50    Loss: 3.586084420979023
Epoch:   11/50    Loss: 

Epoch:   19/50    Loss: 3.5779914759099483
Epoch:   19/50    Loss: 3.5628107637166977
Epoch:   19/50    Loss: 3.5956626545637844
Epoch:   19/50    Loss: 3.5845903187990187
Epoch:   20/50    Loss: 3.5329413836833363
Epoch:   20/50    Loss: 3.460640912875533
Epoch:   20/50    Loss: 3.4283778104931115
Epoch:   20/50    Loss: 3.497613925859332
Epoch:   20/50    Loss: 3.478059810772538
Epoch:   20/50    Loss: 3.4787918657064436
Epoch:   20/50    Loss: 3.4987674567848446
Epoch:   20/50    Loss: 3.5111670274287463
Epoch:   20/50    Loss: 3.498129206895828
Epoch:   20/50    Loss: 3.5430126149207353
Epoch:   20/50    Loss: 3.5041169047355654
Epoch:   20/50    Loss: 3.534894671663642
Epoch:   20/50    Loss: 3.5465824332088234
Epoch:   20/50    Loss: 3.524797761812806
Epoch:   20/50    Loss: 3.5783122550696134
Epoch:   20/50    Loss: 3.569528191164136
Epoch:   20/50    Loss: 3.5687568575143813
Epoch:   20/50    Loss: 3.563305765390396
Epoch:   20/50    Loss: 3.5801593482494356
Epoch:   20/50    L

Epoch:   29/50    Loss: 3.4602421328425406
Epoch:   29/50    Loss: 3.4507012385874987
Epoch:   29/50    Loss: 3.4664278980344534
Epoch:   29/50    Loss: 3.47805983312428
Epoch:   29/50    Loss: 3.4879373222589494
Epoch:   29/50    Loss: 3.51398036070168
Epoch:   29/50    Loss: 3.4810117170214654
Epoch:   29/50    Loss: 3.509173918515444
Epoch:   29/50    Loss: 3.526879256591201
Epoch:   29/50    Loss: 3.5565682377666237
Epoch:   29/50    Loss: 3.5051506206393244
Epoch:   30/50    Loss: 3.451103209611928
Epoch:   30/50    Loss: 3.4125453766435383
Epoch:   30/50    Loss: 3.4089241210371255
Epoch:   30/50    Loss: 3.4200252071022987
Epoch:   30/50    Loss: 3.42307107411325
Epoch:   30/50    Loss: 3.418159082531929
Epoch:   30/50    Loss: 3.4329465851187706
Epoch:   30/50    Loss: 3.4258306436240673
Epoch:   30/50    Loss: 3.465380829572678
Epoch:   30/50    Loss: 3.466207849606872
Epoch:   30/50    Loss: 3.4649736247956753
Epoch:   30/50    Loss: 3.4819396939128637
Epoch:   30/50    Loss:

Epoch:   39/50    Loss: 3.4063506849110126
Epoch:   39/50    Loss: 3.4048154637217523
Epoch:   39/50    Loss: 3.3787527929991485
Epoch:   39/50    Loss: 3.3811198219656946
Epoch:   39/50    Loss: 3.4134032875299454
Epoch:   39/50    Loss: 3.4374142304062842
Epoch:   39/50    Loss: 3.415690538287163
Epoch:   39/50    Loss: 3.4188055228441954
Epoch:   39/50    Loss: 3.436055361106992
Epoch:   39/50    Loss: 3.4275345247238875
Epoch:   39/50    Loss: 3.4354604069143533
Epoch:   39/50    Loss: 3.4172528091818095
Epoch:   39/50    Loss: 3.462311327084899
Epoch:   39/50    Loss: 3.460224222391844
Epoch:   39/50    Loss: 3.5100168213248253
Epoch:   39/50    Loss: 3.4646631266921757
Epoch:   39/50    Loss: 3.477835263684392
Epoch:   39/50    Loss: 3.4807526495307686
Epoch:   39/50    Loss: 3.4723823226988317
Epoch:   40/50    Loss: 3.434191431258204
Epoch:   40/50    Loss: 3.3762931685894726
Epoch:   40/50    Loss: 3.373029573634267
Epoch:   40/50    Loss: 3.397702908143401
Epoch:   40/50    L

Epoch:   48/50    Loss: 3.4417198702692984
Epoch:   48/50    Loss: 3.434024224802852
Epoch:   48/50    Loss: 3.4437305621802805
Epoch:   48/50    Loss: 3.488880232349038
Epoch:   48/50    Loss: 3.482677923515439
Epoch:   48/50    Loss: 3.481697641313076
Epoch:   48/50    Loss: 3.485904024168849
Epoch:   49/50    Loss: 3.4106633298955957
Epoch:   49/50    Loss: 3.3641913484781982
Epoch:   49/50    Loss: 3.3948346238583325
Epoch:   49/50    Loss: 3.379910038411617
Epoch:   49/50    Loss: 3.386973330006003
Epoch:   49/50    Loss: 3.368972373008728
Epoch:   49/50    Loss: 3.406792600825429
Epoch:   49/50    Loss: 3.420383460447192
Epoch:   49/50    Loss: 3.4195685621351
Epoch:   49/50    Loss: 3.3821644365787504
Epoch:   49/50    Loss: 3.4479808609932663
Epoch:   49/50    Loss: 3.4452414851635695
Epoch:   49/50    Loss: 3.438982941955328
Epoch:   49/50    Loss: 3.4280113454908134
Epoch:   49/50    Loss: 3.44359396956861
Epoch:   49/50    Loss: 3.464658565074205
Epoch:   49/50    Loss: 3.45

### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)

For sequence_length, I tried 10, 16, 20 and 10 was better than others.  
And I guessed smaller numbers than 10 might be too small to estimate the next word appropreately from the sequence.

For batch_size,I tried only 64.  
I confirmed that the training was going well.  
In addition, according to the nvidia-smi command, the memory usage was 3333MiB / 7973MiB. I thought it was appropreate.

For num_epochs, I tried 10, 20, 30 and 50.  
30 might be appropreate. But I couldn't get the loss lower than 3.0 at the end of the training.  
So I chose 50 instead of 30.

For learning_rate, I tried only 0.001.  
I found some article that mentioned 0.001 or around numbers were appropreate.

For embedding_dim, I tried 128, 256, 400, and 512.  
At the final, I chose 256 considering that the number of unique words were 46367.  
Because in the Sentiment_RNN_Exercise exsample, we used 400 for 74072 unique words.

For hidden_dim, I tried 128, 256, 400, and 800.  
At the final, I chose 800 because I couldn't get the loss less than 3.5 using smaller numbers as hidden_dim.

For n_layers, I chose 2 because it must be 1-3 from the criteria.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [89]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [90]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        if(train_on_gpu):
            current_seq = current_seq.cpu() # move to cpu
        # the generated word becomes the next "current sequence" and the cycle can continue
        if train_on_gpu:
            current_seq = current_seq.cpu()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [97]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry: i think you're a comedian, i don't want to know how you could make a mistake. you don't even have to pay me to be able to make it up.

elaine:(laughing) i know, it's just the damnedest side.

elaine: what is that?

george:(to himself, picks up his glasses) oh yeah.(kramer walks into the kitchen)

kramer: yeah

jerry:(to elaine) i thought you'd be upset..

jerry: i mean, i don't even have it.(to elaine) so, you want a new george:(pauses. elaine enters)

george: you want to take your shoes?

jerry:(to kramer) oh, i can't believe i just have to be able to get a new suit. i have to tell you, i was thinking of myself, and i was a little rough of the.

george: oh, yeah.

jerry: so, what do you think?

george: well, you know, i think i could take care of you.(jerry leaves)

jerry: hey! hey! hey! i thought you got any good friends.

george: you don't think i could have a ticket, but i don't know how it is.. i don't know..(he pushes jerry in the kitchen; he falls towards his head.)

elai

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [98]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.